src/components/NostoCampaign.ts

```diff
@@ -41,17 +45,12 @@ export async function loadCampaign(element: NostoCampaign) {
 
   const result = await request.load()
   const rec = result.recommendations[element.placement!]
-  const html =
-    typeof rec === "string"
-      ? rec
-      : typeof rec === "object" && rec !== null && "html" in rec
-        ? (rec as { html: string }).html
-        : undefined
-
-  if (html) {
-    // TODO: this is missing attribution handling and HTML preprocessing which will be added later
+  if (element.template && typeof rec === "object" && rec !== null && "products" in rec) {
+    element.toggleAttribute("loading", true)
+    const html = await evaluate(element.template, { recommendation: rec, data: element.dataset })
```

Comment:
in this case I would ignore the dataset and expose recommendation directly as the context